### ISSUE TITLE
Request header field access-control-allow-origin is not allowed by Access-Control-Allow-Headers in preflight response.

### ISSUE THREAD

**Issue by Namrataijare** [aws/serverless-application-model#1887]
Hi,
I have created aws lambda empty function in .netcore3.1. I have created api gateway and enabled the aws lambda proxy integration and also enabled the CORS in api gateway.

Below is my lambda function handler code:

public APIGatewayProxyResponse FunctionHandler(APIGatewayProxyRequest input, ILambdaContext context)
        {
            EmailBody emailBody = new EmailBody();
            emailBody = JsonConvert.DeserializeObject<EmailBody>(input?.Body);

            var response = new APIGatewayProxyResponse
            {
                StatusCode = (int)HttpStatusCode.OK,
                Body = "Mail sent successfully!",
                Headers = new Dictionary<string,
               string> { {"Content-Type", "application/json" } }
            };

            return response;
        }

    public class EmailBody
    {
        public string EmailAddress { get; set; }

        public FilesData[] Files { get; set; }
    }

    public class FilesData
    {
        public string Key { get; set; }
        public string Value { get; set; }
    }
    
    
    I am trying to call the api gateway from but getting below error:

Access to XMLHttpRequest at 'APIGATEWAYURL' from origin 'null' has been blocked by CORS policy: Request header field access-control-allow-origin is not allowed by Access-Control-Allow-Headers in preflight response.

Here is my jquery code:
function SendEmail(){
	debugger;
	var ReceiverEmail = $("#UserEmailId").val().trim();
	var EmailBody = "{\"EmailAddress\":\"XYZ@gmail.com\",\"Files\":[{\"Key\":\"First\",\"Value\":\"HI\"},{\"Key\":\"Second\",\"Value\":\"Hello\"}]}"
	
	$.ajax({
    url: 'APIGATEWAYURL',
    type: 'POST',
    crossDomain: true,
	headers: {'Content-Type': 'application/json','Access-Control-Allow-Origin':'*'},
    data: JSON.stringify(EmailBody),
    dataType: 'json',
    success: function(data) {
		debugger;
        //success stuff. data here is the response, not your original data
    },
    error: function(xhr, ajaxOptions, thrownError) {
		debugger;
        //error handling stuff
    }
});
}

can somebody please help me with this issue?

Thanks,
Namrata

**Comment by chrisoverzero** [comment]
`Access-Control-Allow-Origin` [is not a _request_ header][acao]. You should not be including it in your JQuery request code.

You may need to add CORS headers to your application’s response, as well. I tend to consult this flowchart to remember which headers my services should return in response to the “actual request”.

![](https://www.html5rocks.com/static/images/cors_server_flowchart.png)

Unless you think further problems lie in SAM, someplace like Stack Overflow may be a more appropriate venue for this kind of question. 😺 

[acao]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin

**Comment by Namrataijare** [comment]
Hi @chrisoverzero, 
I tried adding it in application response as well but throwing same error.

**Comment by hoffa** [comment]
The `Access-Control-Allow-Headers` header from the server (during the preflight request) indicates which headers can be sent to the server during the actual request. The error is saying `Access-Control-Allow-Origin` is not part of the allowed headers, and the request fails.

As @chrisoverzero mentioned, `Access-Control-Allow-Origin` should not even be sent during the request; it's a response header sent by the server. If you're having CORS issues making requests, you should remove that header from the request, and check the correct CORS headers are sent by the server (`Access-Control-Allow-Origin`, `Access-Control-Allow-Methods`, `Access-Control-Allow-Headers`, etc.). A cursory search gives some StackOverflow posts with people having similar issues making CORS requests with jQuery (e.g. [this](https://stackoverflow.com/questions/5584923/a-cors-post-request-works-from-plain-javascript-but-why-not-with-jquery) and [this](https://stackoverflow.com/questions/5750696/how-to-get-a-cross-origin-resource-sharing-cors-post-request-working)).

Since this isn't directly related to SAM, but instead a CORS/jQuery issue, I'll close this issue, but feel free to reopen if you think the issues lies with SAM.
